### PR TITLE
fix: use the router API for RouterPreference.PRICE

### PR DESCRIPTION
**Pull request by just-toby**
updates the logic in `src/state/routing/slice.ts` to allow `RouterPreference.PRICE` requests to go through the Router API. There is no need for these to be client-only afaik.

this fixes a bug where the native token price isn't appearing (because we're using "ETH" as the token address).

the follow up is to fix the client-side router to support native tokens.

all other code changes in this PR are just boilerplate to set up a context for the `routerUrl`, so we don't have to pass it around as a prop.

### CODE REVIEW

File: src/hooks/swap/useRouterUrl.tsx [hunk 1]
```diff
@@ -0,0 +1,11 @@
+import { createContext, PropsWithChildren, useContext } from 'react'
+
+const Context = createContext<string | undefined>(undefined)
+
+export function RouterUrlProvider({ children, routerUrl }: PropsWithChildren<{ routerUrl?: string }>) {
```
Comment (tinaszheng) [inline src/hooks/swap/useRouterUrl.tsx:5]: we already have the `swapRouterUrlAtom` so you could use that instead of doing the context

seems more consistent bcuz we store other main props in atoms ^

File: src/state/routing/slice.ts [hunk 2]
```diff
@@ -33,7 +33,7 @@ export const routing = createApi({
 
         if (
           // If enabled, try the routing API, falling back to client-side routing.
-          args.routerPreference === RouterPreference.API &&
+          (args.routerPreference === RouterPreference.API || args.routerPreference === RouterPreference.PRICE_API) &&
```
Comment (tinaszheng) [inline src/state/routing/slice.ts:36]: thoughts on separating the concept of `price` query into its own argument? i feel like conflating routerPreference for both routerUrl and also use case is kinda weird

we could have `routerPreference` just reference either API | Client (and other | Gouda)

and then have a separate argument called quote type - `SwapQuote | PriceQuote | skip` and the polling interval can be determined by just that

**Review by tinaszheng — CHANGES_REQUESTED** [review]
we should probs separate routerPreference vs quote type, but lmk if u disagree or is out of scope for this PR

File: src/state/routing/slice.ts (hunk 2 above)
Comment (just-toby) [reply src/state/routing/slice.ts:36]: yeah that's fair, let me try again

File: src/hooks/routing/types.ts [hunk 3]
```diff
@@ -1,10 +1,19 @@
...
+export enum QuoteType {
+  TRADE = 'trade',
+  PRICE = 'price',
+  SKIP = 'skip',
+}
+
+export type QuoteConfig =
+  | { type: QuoteType.TRADE; preference: RouterPreference }
+  | { type: QuoteType.PRICE; preference: RouterPreference }
+  | { type: QuoteType.SKIP; preference?: undefined }
```
Comment (zzmp) [inline src/hooks/routing/types.ts:15]: Can this just be `| { type: QuoteType.SKIP }`? Why include the (double-)undefined `preference`?

File: src/hooks/routing/types.ts [hunk 4]
```diff
@@ -1,10 +1,19 @@
 export enum RouterPreference {
-  PRICE,
-  SKIP,
   API,
```
Comment (zzmp) [inline src/hooks/routing/types.ts:2]: I really like how you used string-equivalents for the new `enum QuoteType`. It eliminates an ambiguity of default values because `0` is falsey, so you can't check for `RouterPreference | undefined` with a boolean check.

Would you add explicit string values for `enum RouterPreference` too?

File: src/hooks/routing/useRouterTrade.ts [hunk 5]
```diff
@@ -29,32 +30,39 @@ export function useRouterTrade(
...
   const queryArgs = useGetQuoteArgs(
-    { provider, tradeType, amountSpecified, currencyIn, currencyOut, routerPreference, routerUrl },
-    /*skip=*/ routerPreference === RouterPreference.SKIP
+    {
+      provider,
+      tradeType,
+      amountSpecified,
+      currencyIn,
+      currencyOut,
+      routerPreference: quoteConfig.preference,
```
Comment (zzmp) [inline src/hooks/routing/useRouterTrade.ts:48]: At this point, could you just pass `quoteConfig` as the second arg (instead of passing `routerPreference` or `skip`), to keep it more semantic?

File: src/hooks/routing/useRouterTrade.ts [hunk 6]
```diff
@@ -29,32 +30,39 @@ export function useRouterTrade(
...
-  routerPreference: RouterPreference,
-  routerUrl?: string
+  quoteConfig: QuoteConfig
 ): {
   state: TradeState
   trade?: InterfaceTrade
   gasUseEstimateUSD?: CurrencyAmount<Token>
 } {
   const { provider } = useWeb3React()
+  const routerUrl = useAtomValue(swapRouterUrlAtom)
```
Comment (zzmp) [inline src/hooks/routing/useRouterTrade.ts:40]: This can be moved to `useGetQuoteArgs`

**Review by zzmp — APPROVED** [review]
I have small changes requested, but approved assuming those changes are made.
I'm happy to look again after you make them if you'd like.
